pkg/quarterdeck/db/tokens.go

```diff
@@ -57,29 +52,76 @@ func (t *VerificationToken) Sign() (_ string, secret []byte, err error) {
 		return "", nil, err
 	}
 
-	// Compute hash with a random 64 byte key
-	key := make([]byte, keyLength)
-	if _, err = rand.Read(key); err != nil {
-		return "", nil, err
+	return t.signData(data)
+}
+
+// Verify checks that a token was signed with the secret and is not expired.
+func (t *VerificationToken) Verify(signature string, secret []byte) (err error) {
+	if t.Email == "" {
+		return ErrTokenMissingEmail
 	}
 
-	mac := hmac.New(sha256.New, key)
-	if _, err = mac.Write(data); err != nil {
-		return "", nil, err
+	if t.IsExpired() {
+		return ErrTokenExpired
 	}
 
-	// Include the nonce with the key so that the token can be reconstructed later
-	secret = make([]byte, nonceLength+keyLength)
-	copy(secret[0:nonceLength], t.Nonce)
-	copy(secret[nonceLength:], key)
+	if len(secret) != nonceLength+keyLength {
+		return ErrInvalidSecret
+	}
 
-	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil)), secret, nil
+	// Serialize the struct with the nonce from the secret
+	t.Nonce = secret[0:nonceLength]
+	var data []byte
+	if data, err = msgpack.Marshal(t); err != nil {
+		return err
+	}
+
+	return t.verifyData(data, signature, secret)
+}
+
+// NewResetToken creates a token struct from a user ID that expires in 15 minutes.
+func NewResetToken(id ulid.ULID) (token *ResetToken, err error) {
+	if ulids.IsZero(id) {
+		return nil, ErrMissingUserID
+	}
+
+	token = &ResetToken{
+		UserID: id,
+	}
+
+	if token.SigningInfo, err = NewSigningInfo(time.Minute * 15); err != nil {
+		return nil, err
+	}
+
+	if _, err = rand.Read(token.Nonce); err != nil {
```

Comment:
I am curious how this works because `Nonce` is not a direct attribute on `ResetToken`.  Also, is this check redundant since this check is being done within `NewSigningInfo`?

Reply:
Nice catch, I didn't mean to be setting that field twice! The `Nonce` can be accessed like this if you use an embedded struct - note that `SigningInfo` is not specified with a name so `ResetToken` will automatically have all the fields and methods defined on `SingingInfo`. It's a useful pattern when you have very similar structs or structs that have similar functionality. Here's an example: https://gobyexample.com/struct-embedding

Reply:
ah - thanks for providing an example!